addons/product/models/product.py

```diff
@@ -567,6 +567,15 @@ def view_header_get(self, view_id, view_type):
             )
         return super().view_header_get(view_id, view_type)
 
+    @api.constrains('barcode')
+    def _check_barcode_uniqueness(self):
+        """ With GS1 nomenclature, products and packagins use the same pattern. Therefore, we need
+        to ensure the uniqueness between products' barcodes and packagings' ones"""
+        condition = expression.OR([[('barcode', '=', b)] for b in self.mapped('barcode') if b])
+        count = self.env['product.packaging'].search_count(condition)
```

Comment:
Since we only care if at least 1 exists then a `.search(domain, limit=1)` is more efficient in this case (i.e. SQL will stop as soon as it finds 1 matching row vs checking all rows for the condition to determine the total count)